File: YoutubeLinks.Shared/Features/Links/Commands/UpdateLink.cs
using FluentValidation;
using MediatR;
using Microsoft.Extensions.Localization;
using YoutubeLinks.Shared.Features.Links.Helpers;
using YoutubeLinks.Shared.Localization;

namespace YoutubeLinks.Shared.Features.Links.Commands
{
    public static class UpdateLink
    {
        public class Command : IRequest<Unit>
        {
            public int Id { get; set; }
            public string Url { get; set; }
            public string Title { get; set; }
            public bool Downloaded { get; set; }
        }

        public class Validator : AbstractValidator<Command>
        {
            public Validator(IStringLocalizer<ValidationMessage> localizer)
            {
                RuleFor(x => x.Url)
                    .NotEmpty()
                    .WithMessage(x => localizer[nameof(ValidationMessageString.UrlNotEmpty)])
                    .Matches(ValidationConsts.YoutubeVideoUrlRegex)
                    .WithMessage(x => localizer[nameof(ValidationMessageString.VideoUrlMatchesRegex)]);

                When(x => !string.IsNullOrWhiteSpace(x.Title), () =>
                {
                    RuleFor(x => x.Title)
                        .Must(YoutubeHelpers.HaveValidCharactersInTitle)
                        .WithMessage(x => localizer[nameof(ValidationMessageString.TitleHaveValidCharacters)])
                        .MaximumLength(YoutubeHelpers.MaximumTitleLength)
                        .WithMessage(x => localizer[nameof(ValidationMessageString.TitleMaximumLength)]);
                });
            }
        }
    }
}

File: youtube-links/src/shared/features/links/commands/UpdateLink.ts
import { ValidationConsts } from '@/shared/localization/ValidationConsts';
import { YoutubeHelpers } from '../helpers/YoutubeHelpers';

export namespace UpdateLink {
  export interface Command {
    id: number;
    url: string;
    title: string;
    downloaded: boolean;
  }

  export const Validation = {
    url: {
      notEmpty: (v: string) => !!v || 'Youtube video url should not be empty.',
      matchesYoutubeVideoUrlRegex: (v: string) =>
        ValidationConsts.matchesYoutubeVideoUrlRegex(v) ||
        'This is not a valid link to the YouTube video.',
    },
    title: {
      mustHaveValidCharactersInTitle: (v: string) =>
        !v.trim() ||
        YoutubeHelpers.haveValidCharactersInTitle(v) ||
        'Title contains invalid characters.',
      maximumLength: (v: string) =>
        v.length <= ValidationConsts.MaximumStringLength ||
        `The length of title must be ${ValidationConsts.MaximumStringLength} characters or fewer. You entered ${v.length} characters.`,
    },
  };
}
